ts/kpt-functions/src/types.ts

```diff
@@ -274,6 +274,72 @@ export function kubernetesKey(o: KubernetesObject): string {
   return `${o.apiVersion}/${o.kind}/${namespace}/${o.metadata.name}`;
 }
 
+/**
+ * ResourceList is the wire format for the output of the KPT function as defined by the spec:
+ * https://github.com/kubernetes-sigs/kustomize/blob/master/cmd/config/docs/api-conventions/functions-spec.md
+ */
+export class ResourceList implements KubernetesObject {
+  readonly apiVersion = 'v1';
+  readonly kind = 'ResourceList';
+  readonly metadata = {
+    name: 'output',
+  };
+  readonly items: KubernetesObject[];
+  readonly issues?: Issue[];
+
+  /**
+   * @param items List of Kubernetes objects returned by the function.
+   * @param issues List of issues returned by the function.
+   */
+  constructor(items: KubernetesObject[], issues?: Issue[]) {
+    this.items = items;
+    this.issues = issues;
+  }
+}
+
+/**
+ * Severity of a configuration issue.
+ */
+type Severity = 'error' | 'warning' | 'note';
+
+/**
+ * Issue represents a configuration-related issue returned by a function.
+ * 
+ * It can be at the following granularities:
+ * - A file containing multiple objects
+ * - A specific kubernetes object
+ * - A specific field of a kubernetes object
+ */
+export interface Issue {
+  // Severify of the issue.
+  severity: Severity;
+  // Message describing the issue.
+  message: string;
+  // Additional metadata for tracking the issue.
+  tags?: { [key: string]: string };
+  // A reference to an object with the issue.
+  resourceRef?: {
+    apiVersion: string;
+    kind: string;
+    namespace: string;
+    name: string;
+  }
+  // File-level for the issue.
+  file?: {
+    // OS agnostic, relative, slash-delimited path.
+    // e.g. some-dir/some-file.yaml
+    path: string;
+    // Index of the object in a multi-object YAML file.
+    index: number;
+  }
+  // A specific field in the object.
+  field?: {
+    // JSON Path (e.g. spec.template.spec.containers[3].resources.limits.cpu)
+    path: string;
```

Comment:
suggested / current value

Reply:
Done. Initially left this out caused they seemed like advanced feature we can add in the future. Any object with making this `string | number | boolean` so the function implementor doesn't need to cast primitives to string?